File: Go/1 Arrays/Q5.go
// In this program, we will enter two sorted arrays. While merging them, we will compare the elements of both the arrays and merge them in a sorted manner.

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)

	fmt.Println("Enter the size of the first array:")
	scanner.Scan()
	n1, _ := strconv.Atoi(scanner.Text())
	a := make([]int, n1)
	fmt.Println("Enter the elements of the first array:")
	for i := 0; i < n1; i++ {
		scanner.Scan()
		a[i], _ = strconv.Atoi(scanner.Text())
	}

	fmt.Println("Enter the size of the second array:")
	scanner.Scan()
	n2, _ := strconv.Atoi(scanner.Text())
	b := make([]int, n2)
	fmt.Println("Enter the elements of the second array:")
	for i := 0; i < n2; i++ {
		scanner.Scan()
		b[i], _ = strconv.Atoi(scanner.Text())
	}

	c := mergeSortedArrays(a, b)

	fmt.Println("Final array after merging:")
	for _, value := range c {
		fmt.Printf("%d ", value)
	}
	fmt.Println()
}

func mergeSortedArrays(a, b []int) []int {
	n1, n2 := len(a), len(b)
	c := make([]int, 0, n1+n2)

i, j := 0, 0

	for i < n1 && j < n2 {
		if a[i] < b[j] {
			c = append(c, a[i])
			i++
		} else {
			c = append(c, b[j])
			j++
		}
	}

	c = append(c, a[i:]...)
	c = append(c, b[j:]...)

	return c
}

File: Rust/1 Arrays/Q5.rs
// In this program, we will enter two sorted arrays. While merging them, we will compare the elements of both the arrays and merge them in a sorted manner.


use std::io;

fn main() {
    let mut input = String::new();

    println!("Enter the size of the first array:");
    io::stdin().read_line(&mut input).expect("Failed to read input");
    let n1: usize = input.trim().parse().expect("Please enter a valid number");
    input.clear();

    let mut a = Vec::new();
    println!("Enter the elements of the first array:");
    for _ in 0..n1 {
        io::stdin().read_line(&mut input).expect("Failed to read input");
        let element: i32 = input.trim().parse().expect("Please enter a valid integer");
        a.push(element);
        input.clear();
    }

    println!("Enter the size of the second array:");
    io::stdin().read_line(&mut input).expect("Failed to read input");
    let n2: usize = input.trim().parse().expect("Please enter a valid number");
    input.clear();

    let mut b = Vec::new();
    println!("Enter the elements of the second array:");
    for _ in 0..n2 {
        io::stdin().read_line(&mut input).expect("Failed to read input");
        let element: i32 = input.trim().parse().expect("Please enter a valid integer");
        b.push(element);
        input.clear();
    }

    let mut c = Vec::new();
    let mut i = 0;
    let mut j = 0;

    while i < n1 && j < n2 {
        if a[i] < b[j] {
            c.push(a[i]);
            i += 1;
        } else {
            c.push(b[j]);
            j += 1;
        }
    }

    while i < n1 {
        c.push(a[i]);
        i += 1;
    }

    while j < n2 {
        c.push(b[j]);
        j += 1;
    }

    println!("Final array after merging:");
    for element in &c {
        print!("{} ", element);
    }
    println!();
}